pkg/daemon/certs/certs.go

```diff
@@ -106,69 +150,132 @@ func splitAtPEMCert() func(data []byte, atEOF bool) (advance int, token []byte,
 		windows.CERT_SYSTEM_STORE_LOCAL_MACHINE, uintptr(unsafe.Pointer(syscall.StringToUTF16Ptr("ROOT"))))
 	if err != nil {
-		return false, fmt.Errorf("Failed to open root certificate store: %v", err)
+		return false, fmt.Errorf("Failed to open root certificate store: %w", err)
 	}
 	defer func() {
 		if err := windows.CertCloseStore(systemStore, 0); err != nil {
 			klog.Error("Failed to close root system certificate store")
 		}
 	}()
+	klog.Infof("opened root cert store")
 
-	// Get all existing from system store
-	existingCerts, err := getAllCerts(systemStore)
+	certsImported, err := ensureCertsAreImported(systemStore, expectedCerts, existingCerts)
+	if err != nil {
+		return false, fmt.Errorf("Failed to import certificates: %w", err)
+	}
+	certsDeleted, err := ensureCertsAreDeleted(systemStore, expectedCerts, existingCerts)
 	if err != nil {
-		return false, fmt.Errorf("Failed to read certificate file: %v", err)
+		return false, fmt.Errorf("Failed to delete certificates: %w", err)
 	}
+	return (certsImported || certsDeleted), nil
+}
 
-	certChange := false
+// ensureCertsAreImported imports the expected certificates into the instance's root system trust store, if not already
+// present. Returns a boolean if any certificates were imported
+func ensureCertsAreImported(store windows.Handle, expectedCerts, existingCerts []*x509.Certificate) (bool, error) {
+	certImported := false
 	for _, cert := range expectedCerts {
 		if containsCert(existingCerts, cert) {
 			// Cert already exists as expected, do nothing
 			continue
 		}
 
 		// Add the certificate to the instance's system-wide trust store
-		if err := addCertToStore(systemStore, cert); err != nil {
+		if err := addCertToStore(store, cert); err != nil {
 			return false, err
 		}
-		certChange = true
+		certImported = true
 	}
-	return certChange, nil
+	return certImported, nil
 }
 
-// getAllCerts reads all the certs from the given certificate store
-func getAllCerts(store windows.Handle) ([]*x509.Certificate, error) {
-	var existingSystemCerts []*x509.Certificate
-	var certContext *windows.CertContext
-	var err error
-	for {
-		certContext, err = windows.CertEnumCertificatesInStore(store, certContext)
-		if err != nil {
-			if errors.Is(err, windows.Errno(windows.CRYPT_E_NOT_FOUND)) {
-				// Error code implies we have read all certs:
-				// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certenumcertificatesinstore
-				break
+// ensureCertsAreDeleted deletes any existing certificates that are not in the expected list.
+// Returns a boolean if any certificates were deleted.
+func ensureCertsAreDeleted(store windows.Handle, expectedCerts, existingCerts []*x509.Certificate) (bool, error) {
+	certDeleted := false
+	for _, cert := range existingCerts {
+		// Any imported previously certs that are NOT in the new cert bundle should be removed
+		if !containsCert(expectedCerts, cert) {
+			if err := removeCertFromStore(store, cert); err != nil {
+				return false, err
```

Comment:
Shouldn't we return `certDeleted` here? I am thinking about the case where there are multiple certs that need to be removed but we error on deleting only one. Or will we reconcile again on error? Same question holds for `ensureCertsAreImported()`

Reply:
We will reconcile again on error. This is consistent with the approach for env vars -- if we fail to read/set/remove one of them, we return error and try the transaction again